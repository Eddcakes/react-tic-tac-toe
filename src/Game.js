import React, {useState} from "react";
import OrderButton from "./components/OrderButton";
import Board from "./components/Board";
import {calculateCoord, calculateWinner} from "./utils/utils";

function Game({size, ...props}) {
  const [history, setHistory] = useState([Array(9).fill(null)]);
  const [stepNumber, setStepNumber] = useState(0);
  const [xIsNext, setXIsNext] = useState(true);
  const [coordinate, setCoordinate] = useState([Array(2).fill(null)])
  const [isAssending,setIsAssending] = useState(true)
  const current = history[stepNumber];
  const winner = calculateWinner(current);
  const handleClick = i => {
    if (winner || current[i]) {
      //if winner or same square then do not allow anymore play
      return;
    }
    const curBoard = current.slice();
    curBoard[i] = xIsNext ? "X" : "O";
    setHistory(history.slice(0, stepNumber + 1).concat([curBoard]));
    setStepNumber(stepNumber + 1);
    setXIsNext(!xIsNext);
    let lastBoard = history[history.length - 1]
    let theMove = curBoard.map( (value, i) => {
      //find difference between current board and last board if different save else replace with null
      curBoard[i] === lastBoard[i] ? value = null : value = curBoard[i]
      return value
    })
    setCoordinate([...coordinate.slice(0, stepNumber + 1), calculateCoord(size, theMove)])
  };
  const jumpTo = step => {
    setStepNumber(step);
    //if the number we are changing to is even then x is next
    setXIsNext(step % 2 === 0);
  };
  const moves = history.map((step, move) => {
    const btnDesc = move ? `Go to move #${move} (${coordinate[move]})` : `Go to game start`;
    let currentStyle;
    stepNumber === move
      ? (currentStyle = { fontWeight: "bold" })
      : (currentStyle = { fontWeight: "normal" });
    return (
      <li key={"GoTo" + move}>
        <button onClick={() => jumpTo(move)} style={currentStyle}>{btnDesc}</button>
      </li>
    );
  });
  let status;
  let boardSqs = document.getElementsByClassName('square')
  //is this the best way to set a style for winner and remove it ?
  if (winner) {
    status = `Winner: ${winner.icon}`;
    winner.winningMove.map( (v) => boardSqs[v].style.background = 'yellow' )
  }else if (moves.length > size * size){ //size*size is all the available box's so we know it is draw
    status = `This game is a draw!!!`
  }else{
    status = `Next player: ${xIsNext ? "X" : "O"}`;
    for (let boardSq of boardSqs){
      boardSq.style.background = 'none'
    }
  }
  const reorder = () => {
    setIsAssending(!isAssending)
  }
  let movesByOrder //moves or moves.reverse
  let isReversed //boolen
  if (isAssending){
    movesByOrder = moves
    isReversed = false
  }else{
    movesByOrder = moves.slice().reverse()
    isReversed = true
  }
  return (
    <>
    <h1>React tic-tac-toe with hooks</h1>
    <div className="game">
      <div className="game-board">
        <Board 
          squares={current} 
          onClick={i => handleClick(i)}
          size={size}
        />
      </div>
      <div className="game-info">
        <div>{status}</div>
        <div>
          <OrderButton 
            reorderClick={() => reorder()} 
            btnText={isAssending ? "Set decending" : "Set Assending"}
          />
          </div>
        <ol reversed={isReversed}>{movesByOrder}</ol>
      </div>
    </div>
    </>
  );
}

export default Game