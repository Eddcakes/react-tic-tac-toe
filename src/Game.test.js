import React from "react";
import Game from "./Game";

/*
JEST -
test(name, fn, timeout)
Also under the alias: it(name, fn, timeout)
*/

import {
  render,
  getByText,
  fireEvent,
  cleanup,
  waitForElement,
} from "@testing-library/react"

//jest after each test run cleanup
afterEach(cleanup);

it("renders", () => {
  const { asFragment } = render(<Game size={3}/>);
  //Snapshots aren't always great if we want to change html -> tree u to update tests
  expect(asFragment()).toMatchSnapshot();
});

/*
test("Allows user to set ascending/descending", () => {
  const { getByText } = render(
    <Game size={3}/>
  );
  const orderButton = getByText("set")

  //fireEvent.click(orderButton)
})
*/