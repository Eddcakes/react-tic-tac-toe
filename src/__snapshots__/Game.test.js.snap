// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <h1>
    React tic-tac-toe with hooks
  </h1>
  <div
    class="game"
  >
    <div
      class="game-board"
    >
      <div>
        <div
          class="board-row"
        >
           
        </div>
        <button
          class="square"
          data-testid="zone0"
          id="zone0"
          type="button"
        />
        <button
          class="square"
          data-testid="zone1"
          id="zone1"
          type="button"
        />
        <button
          class="square"
          data-testid="zone2"
          id="zone2"
          type="button"
        />
        <div
          class="board-row"
        >
           
        </div>
        <button
          class="square"
          data-testid="zone3"
          id="zone3"
          type="button"
        />
        <button
          class="square"
          data-testid="zone4"
          id="zone4"
          type="button"
        />
        <button
          class="square"
          data-testid="zone5"
          id="zone5"
          type="button"
        />
        <div
          class="board-row"
        >
           
        </div>
        <button
          class="square"
          data-testid="zone6"
          id="zone6"
          type="button"
        />
        <button
          class="square"
          data-testid="zone7"
          id="zone7"
          type="button"
        />
        <button
          class="square"
          data-testid="zone8"
          id="zone8"
          type="button"
        />
      </div>
    </div>
    <div
      class="game-info"
    >
      <div>
        Next player: X
      </div>
      <div>
        <button>
          Set decending
        </button>
      </div>
      <ol>
        <li>
          <button
            style="font-weight: bold;"
          >
            Go to game start
          </button>
        </li>
      </ol>
    </div>
  </div>
</DocumentFragment>
`;
